File: app/javascript/channels/search_channel.js
import consumer from 'channels/consumer';
import fetchAnalytics from 'analytics/fetch_analytics';

const searchChannel = (ipAddress) =>
  consumer.subscriptions.create(
    { channel: 'SearchChannel', ip_address: ipAddress },
    {
      connected() {
        // Called when the subscription is ready for use on the server
        console.debug(
          `Connected to SearchChannel subscription for ip address ${ipAddress}`
        );
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
        console.debug(
          `Disconnected from SearchChannel subscription for ip address ${ipAddress}`
        );
      },

      sendSearch(query, sessionId) {
        console.log({ sendSearch: this });
        this.perform('receive', {
          query: query,
          ip_address: ipAddress,
          complete: false,
          session_id: sessionId,
        });
      },

      completeSearch(query, sessionId) {
        console.log({ completeSearch: this });
        this.perform('receive', {
          query: query,
          ip_address: ipAddress,
          complete: true,
          session_id: sessionId,
        });
        setTimeout(() => {
          fetchAnalytics();
        }, 500);
      },
    }
  );

export default searchChannel;

File: app/channels/search_channel.rb
class SearchChannel < ApplicationCable::Channel
  def subscribed
    stream_from "search_channel_#{params[:ip_address]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    query = data["query"]
    ip_address = data["ip_address"]
    session_id = data["session_id"]
    complete = data["complete"]

    if complete
      Search.create(session_id:, query:, ip_address:, complete: true)
      # Clear the cache after saving to the database
      Rails.cache.delete(session_id)
    else
      # Replace the previous query with the current one or add a new query in the cache
      Rails.cache.write(session_id, query, expires_in: 10.minutes)
    end
  end
end
